Release 0.28.0 (#413)
ponylang/ponylang-website: content/blog/0.28.0-released.md

+++
draft = false
author = "seantallen"
categories = [
    "Release",
]
date = "2019-03-22T16:06:34-04:00"
title = "0.28.0 Released"
+++
Pony 0.28.0 is a high-priority release. We advise updating as soon as possible.

> In addition, a high-priority bug fix, there are "breaking changes" if you build Pony from source. And, we've dropped support for some Debian and Ubuntu versions. Full details are below...

Reviewer: ```suggestion
In addition to a high-priority bug fix, there are "breaking changes" if you build Pony from source. We've also dropped support for some Debian and Ubuntu versions. Read on for further details.
```

> In addition, a high-priority bug fix, there are "breaking changes" if you build Pony from source. And, we've dropped support for some Debian and Ubuntu versions. Full details are below...

Reviewer 2: This sentence sounds off to me, but maybe that's because I'm not a native speaker. But I'd go with something like:

    In addition to a high-priority bug fix, there are "breaking changes" if you build Pony from source. We've also dropped support for some Debian and Ubuntu versions. Full details are below...

<!--more-->

## Ensure methods reached through a union are detected as reachable

Prior versions of Pony contained a bug in code reachability analysis that caused deterministic runtime crashes due to undetected reachability of a method call under certain specific conditions, including:

- the method being called only through a trait within a union type
- the order of compilation causing the concrete type to be seen as reachable after reaching the method call

The issue has been fixed in this release, and it's recommended that all users update to prevent the possibility of such runtime crashes.

## Stop trying to be clever when finding user's LLVM installation

> Clever is bad. Eventually it bites you in the ass. Ask The Narrator, he

Reviewer: ```suggestion
```

Reviewer: it's superfluous

knows.

Prior to this commit, we were trying to be clever and helpful and would
bend over backwards to try and find the user's LLVM installation
whereever their package manager of choice might put it. This was, in the
end, a losing battle. We were constantly adding more and more
variations.

To make it worse, we tried to have a hierarchy where we would look for
LLVM 7 first then 6 etc. We did this by looking something like:

- llvm-config-7
- llvm-config-6

The problem with this approach was that some installations, wouldn't use
the `-7` notation. For example, if you download a release directly from
releases.llvm.org, then your llvm-config will simply be `llvm-config`.

We discovered this when, after a user installed LLVM 7 from releases.llvm.org,
and the build process continued to have LLVM 6 used because the search loooked something
like:

- llvm-config-7
- llvm-config-6
- ... more stuff ...
- llvm-config

> In the end, that's all just too much clever.

Reviewer: If the prose above is contained in a PR it'd be better to link it here. It'd suffice to say something simple like: `The first llvm-config we find in your path will be used.  To override this behaviour, set LLVM_CONFIG to your preferred LLVM installation.`

Author: It's our policy to provide inline content for major changes.

With this change, we'll use the first `llvm-config` we find in your path
if you haven't set the LLVM_CONFIG environment variable which should be
considered the preferred way to select an LLVM when building ponyc.

## Stop building packages for older Ubuntu and Debian's

> In particular, as of 0.28.0 we no longer build binary ponyc packages for Ubuntu Trusty and Artful as well as Debian Jessie. We are still building binary packages for Ubuntu Xenial and Bionic as well as Debian Stretch and Buster.

Reviewer: ```suggestion
As of 0.28.0 we no longer build binary ponyc packages for Ubuntu Trusty and Artful as well as Debian Jessie. See the supported platforms page <here>.
```
